### ISSUE TITLE
Improve zipping performance

### ISSUE THREAD

**Issue by rylwin** [caxlsx/caxlsx#55]
We're starting to generate some pretty large Excel files with caxlsx and, as our files get larger, so do the resources consumed to generate the file. We're working on ways to decrease the time it takes to render Excel files, part of which is due to slowness in our own app (slow queries, etc) and part of which is due to generating the Excel file.

One area we've found for potentially decreasing the time it takes to generate the Excel file is by using the system `zip` tool instead of rubyzip. We've hacked together a solution that seems to work for us and yield some pretty great improvements.

We have a gist to demonstrate the performance of the "built-in" rubyzip that's currently used by caxlsx vs. the "system-zip" binary: https://gist.github.com/rylwin/013eaa9768e94b75572d34e1695ac472. 

On my personal laptop, the benchmark yields:
```
$ ruby zip-perf.rb
                  user     system      total        real
built-in:     0.503427   0.258035   0.761462 (  0.597083)
system-zip:   0.169251   0.000000   0.202789 (  0.201150)
```

And on a Heroku instance:
```
$ ruby zip-perf.rb
                  user     system      total        real
built-in:     2.760000   2.176000   4.936000 (  4.260702)
system-zip:   0.424000   0.000000   0.452000 (  0.483958)
```

In both cases, the speed up is quite drastic.

If there's interest in incorporating the use of an alternate zip library, I'd be happy to work on a PR. A couple thoughts:

* Shelling out to the system `zip` binary adds a dependency on `zip`, which may not always be there. Perhaps via config the user could select how they want to render the Excel file, or we could check for the presence of `zip`
* I'm on a linux box (I would assume Mac would work fine, not sure about Windows, more exploration would be necessary)
* I tried using an old gem I found called `zipruby` that provides `libzip` bindings, but it's last commit is ~7 years ago and when I load it, it conflicts with `rubyzip`, so while I'd prefer to use something like `zipruby` to avoid shelling out to `zip`, that would be a bigger lift, but perhaps worth it.

So that's where we are as of now. Curious to hear the community's thoughts and whether any of this is of interest.

P.S. Thanks for maintaining this great gem!

**Comment by noniq** [comment]
I think that's a great idea! Maybe this could be an option in `Axlsx::Package#serialize`? Here's a spontaneous idea:

```ruby
package.serialize("test.xlsx", external_zip: true) # uses zip executable in path
package.serialize("test.xlsx", external_zip: "/my/special/zip") # uses specified executable
```

Thoughts?

**Comment by rylwin** [comment]
@noniq I think what you proposed would work well. 

I have a couple other thoughts for consideration, but I'd also be happy to go with what you proposed:

* Thoughts on `#to_stream` also supporting the option?

    It would sort of be a lie to the user because it's having to go use the disk to generate the contents. The advantage of it would be that as the user, I would let `to_stream` figure out where to put the tmp file and to clean it up for me.

    For example, in our use-case, we just want the stream generated as quickly as possible. If only `#serialize` supported the external_zip option, our code would be something like the follow psuedo-code:

    ````
    random_file_name = ...
    package.serialize(random_file_name, external_zip: true)
    contents = File.read(random_file_name)
    `rm random_file_name`
    send_data contents filename: user_friendly_file_name
    ````

* As a user of the gem, would I prefer to pass an option wherever I invoke `#serialize`? Or have a config-like variable that I could set, e.g., `Axlsx.config.zip_provider = :external`

**Comment by noniq** [comment]
Good point about the problem with `#to_stream` supporting this option. Out of curiosity: Wouldn’t it be possible to implement your usecase like this (untested)?

```ruby
Tempfile.open do |tempfile|
  package.serialize(tempfile.path, external_zip: true)
  send_data tempfile.rewind.read # or even `send_file tempfile.path`
end
```

I’m not against a config option per se, but I’m a bit hesitant because so far Axlsx has no configuration at all – not sure if it's worth introducing one for a single option only.

**Comment by rylwin** [comment]
> I’m not against a config option per se, but I’m a bit hesitant because so far Axlsx has no configuration at all – not sure if it's worth introducing one for a single option only.

Makes sense 

> Good point about the problem with `#to_stream` supporting this option. Out of curiosity: Wouldn’t it be possible to implement your usecase like this (untested)?

Good call on `Tempfile`, that would absolutely be better than implementing custom random file logic.

I was wondering what your thoughts would be on implementing something like `#to_stream(external_zip: true)`, as a way for the user to not have to worry about the tempfile at all, but I'm happy to leave `#to_stream` out of this entirely -- it's simple enough for the user to use `#serialize` instead as you show in the above code example.

It sounds like your preference would be to not support `external_zip` via `#to_stream`, which is fine by me. Let me know if I've misunderstood, otherwise, I'll work on a PR that implements the interface you initially suggested:

```ruby
package.serialize("test.xlsx", external_zip: true) # uses zip executable in path
package.serialize("test.xlsx", external_zip: "/my/special/zip") # uses specified executable
```

Thanks for all the input!

**Comment by noniq** [comment]
Regarding `#to_stream` I think you already pointed out the problem that it could be surprising if (temporary) files are created when using this method. But I’m not sure: Maybe Rubyzip creates temporary files anyway? In this case (which probably should be documented) it wouldn’t be very much of a difference to use an external zip program. Would you like to investigate this further?

Or maybe we should add another way of serializing a package to better fit these use-cases. Maybe something like

```ruby
package.serialize do |temporary_package_file|
  # do something with the file
end # temporary package file cleaned up automatically
```

**Comment by straydogstudio** [comment]
Nice work @rylwin. I too am hesitant to start configuration. I like the `external_zip` parameter.

@noniq How do you see your temporary_package_file being used? How would the default case be handled?

**Comment by rylwin** [comment]
Thanks @straydogstudio 

The TL;DR is I think the path forward here is to add the `external_zip` option to the `#serialize` method. I'll start work on this as soon as I have some free time (probably this weekend at the latest).

The full story is that the `temporary_package_file` was @noniq's suggestion to appease my specific use-case (thanks, @noniq!). I have a Rails app where I send the Excel as a response using Rails' `send_data`. Since in my use case, I don't really care to have a file, I was trying to see if there was a way to use `#to_stream` with the proposed `external_zip` option. However, `external_zip` requires the filesystem be used, and the name `#to_stream` might imply no filesystem involvement; we concluded writing to disk in `#to_stream` might be surprising behavior. So `temporary_package_file` was a solution proposed that would allow me to do something like:

```ruby
package.serialize(external_zip: true) do |temporary_package_file|
  send_data File.read(temporary_package_file)
end # temporary package file cleaned up automatically
```

I'd be just as happy with @noniq's original suggestion:

```ruby
Tempfile.open do |tempfile|
  package.serialize(tempfile.path, external_zip: true)
  send_data tempfile.rewind.read # or even `send_file tempfile.path`
end
```

If we decide we like the block approach or don't have concerns about `#to_stream` also accepting the `external_zip` option, I'd be happy to expand the PR to cover either of these. In my use-case it would slightly simplify the code in my Rails app, but it's by no means necessary for me to take advantage of the speedup.

**Comment by noniq** [comment]
👍 

I’d say let’s keep this issue / PR focused on the external zip option for now. 

Reconsidering the name and usage of the option: I just realized that we can probably always require a string – either the full path to a specific zip executable, or simply `"zip"` to use the first zip executable found in the search path. (I imagine we could simply pass the command name to `#system` and let the shell do all lookups and path handling.)

Here’s how it would look like for the case when just using `"zip"`:

```ruby
serialize(external_zip: "zip")
serialize(zip_provider: "zip")
serialize(zip_command: "zip")
serialize(zip_binary: "zip")
serialize(zip_executable: "zip")
serialize(zip: "zip")
```

I think I’d prefer `zip_command: "zip"`, eventhough this somehow implies that arguments can also be specified. But why not? Maybe it would even make sense performancewise to use `zip_command: "zip -1"` (“compress faster”)? Or is this too confusing or hard to implement?

**Comment by rylwin** [comment]
I like it (`serialize(zip_command: "zip")`). I don't think it would be any harder to allow passing options, that's a nice idea

**Comment by straydogstudio** [comment]
Would we ever want a third option besides RubyZip or a command? I don't think so. If not, then zip_command is a good choice.

**Comment by noniq** [comment]
Then let's go for `zip_command`! 🚀
